ARCHITECTURE.md

```diff
@@ -0,0 +1,209 @@
+
+Prefer Ruby/Rails/MySQL on the server side and JS/TS/React on the client side. Prefer Ruby/Rake for general tooling (infra, build, etc) and nodejs for JS/TS-related tooling. Most code.org work should be implemented with these core technologies.
+
+<details>
+  <summary>Why?</summary>
+
+  * We generally have simple "rails shaped", "mysql shaped" and "react shaped" problems that aren't very exotic and don't benefit significantly from exotic solutions.
+  * We have a significant investment in Ruby/Rails that makes it challenging to do a wholesale move to a different platform.
+  * We get productivity benefits from being a monolith, which means sticking with the same basic language + web framework.
+  * Ruby/Rails is working well. While other platforms (e.g., Go) may be more optimized/efficient, they don’t outweigh the effort that would be required to move.
+  * A single Ruby/Rails server-side architecture makes it easier for team members to move between different parts of the code base.
+  * We don't have sections of the code "go dark" when contributors leave the project taking with them the last knowledge of relatively obscure technology.
+  * A good general principle for code.org: given our problem space, we should usually be making pretty boring technology choices.
+
+</details>
+
+<details>
+  <summary>Exceptions</summary>
+
+  * When it is fundamentally impossible to implement a specified feature without using another language, e.g. Javabuilder uses Java because our curriculum goal required running Java code, which inherently requires Java.
+  * If it is not feasible to implement a feature in Ruby due to lack of libraries, Python may be invoked from ActiveJob ([howto](https://github.com/code-dot-org/code-dot-org/tree/staging/python)).
+  * May use *short* snippets of inline JS/JQuery *without* React on the client side, where pragmatic.
+  * May use bash or nodejs for general tooling when pragmatic (e.g. better nodejs apis than ruby)
+  * Consider Redis instead of MySQL when performance is important and ephemerality is acceptable (e.g. cache values)
+  * This tenet will probably have more exceptions and edge cases than most. Its not an exhaustive list, its a "highly preferred list".
+    * Be pragmatic, do a good faith attempt to use preferred languages and frameworks, but solve your problems.
+    * E.g. we also use cloudformation (etc etc) which could be argued to be kind of like a programming language, but is **totally** not a problem.
+      * OTOH, you don't get to "use Haskell because everything else should be considered lame and fundamentally broken" 😉.
+
+</details>
+
+### Tenet 5: “The Rails Way” (vs. overfitting to today’s needs)
+
+Prefer the conventional Rails way to accomplish a task: it might be better to be 100% conventional than get a 10% gain by deviating from convention. This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.
```

Comment:
as a litmus test, should someone be able to use this tenet to support PR feedback saying that a new ajax route associated with reading or writing specific models should be added to regular controllers like controllers/sections_controller, as opposed to our current alternative of using controllers/api/v1/sections_controller? I haven't done an accurate count but I think we use api/v1 about 65% of the time today. 

For example, if the following controller action(s) were being added today, should it be moved to app/controllers? https://github.com/code-dot-org/code-dot-org/blob/f1492ffb32e49b95a61fdc60d19dde7c415132a6/dashboard/app/controllers/api/v1/amazon_future_engineer_controller.rb#L13

Reply:
meta-comment: of all the tenets, i'm concerned that this one leave the most ambiguity around when to follow it vs follow the most common practice in our codebase.

Reply:
I agree this tenet is the most ambiguous, but that's also why its powerful (it tracks lots of specific upstream decisions going forward without having to specify them).

That said I was hoping this language makes it clear that unlike many of the tenets specific as SHOULD or MUST, this one is "PREFER the conventional Rails way". And additionally says: "This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.

That means this tenet can't make the decision for you, but yes, you should think twice before starting a path like v1. In fact, I think v1 is a **great** example of "why did we do this weird semi-pointless deviation from rails instead of sticking with rails stock?!?".

Reply:
> should someone be able to use this tenet to support PR feedback saying that a new ajax route

imo probably yes, at least following the "think twice" the tenet advices.

> should it be used to support PR feedback saying "this weird v1 thing in the first place?"

emphatically YES, that's a great example where deviating from stock rails proved (afaik) to be weird and pointless, sorta following a fad at the time, and has created future questions like "should someone be able to use this tenet to support PR feedback saying that a new ajax route" due to it being different from stock rails.

Reply:
Thank you, the emphasis on PREFER is helpful.

FWIW there was a real gain for using `api/v1` routes initially -- I believe it was used to simplify the logic for making those routes available to pegasus. However, I agree that api/v1 route prefix no longer appears to serve any purpose. Furthermore, it frequently causes headaches when similar routes are split between multiple controllers with subtly different names and paths.

Reply:
Oh interesting, always nice to learn history for codebases, really helps understanding